### PR TITLE
Add zypper package manager retries

### PR DESCRIPTION
**Pull request by benank**
- Add retries for all invocations to the package manager
- Ensure ZYPP_LOCK_TIMEOUT environment variable is set so it will be used
- Update version to 1.6.28

### CODE REVIEW

**Comment by benank** [comment]
Still working on tests.

Should I remove the retry code in `refresh_repo` and keep the reboot logic?

File: src/core/src/package_managers/ZypperPackageManager.py
Comment (kjohn-msft) [inline src/core/src/package_managers/ZypperPackageManager.py:55]: Read what the original value is first (and log). Do all of this as you enter invoke_package_manager and reset to original value when you exit from the function. Setting env. var is a low cost operation so the repeats are okay. We shouldn't leave permanent configuration changes on customer machines. There is still a possibility of failure where the value is not reset, but we can address that case later.

Comment (benank) [reply src/core/src/package_managers/ZypperPackageManager.py]: Done. Split into get/set and moved to invoke_package_manager

File: src/core/src/bootstrap/EnvLayer.py
Comment (kjohn-msft) [inline src/core/src/bootstrap/EnvLayer.py:60]: We try to be OS agnostic in the environment layer - everything that is not, gets passed in usually. No need to block the PR on this; just FYI.

File: src/core/src/bootstrap/EnvLayer.py
Comment (kjohn-msft) [inline src/core/src/bootstrap/EnvLayer.py:127]: This is best effort. 

Do not raise as things can succeed even without this. But because this is a shared class, define the behavior on whether to raise or not through a default variable in the function signature (see examples of 'raise_if_not_found' / similar elsewhere)

File: src/core/src/bootstrap/EnvLayer.py
Comment (kjohn-msft) [inline src/core/src/bootstrap/EnvLayer.py:155]: Same comment about when to raise.

File: src/core/src/package_managers/ZypperPackageManager.py
Comment (kjohn-msft) [inline src/core/src/package_managers/ZypperPackageManager.py:95]: zypp_lock_timeout_backup

**Review by kjohn-msft — CHANGES_REQUESTED** [review]
Comments inline

File: src/core/src/package_managers/ZypperPackageManager.py
Comment (benank) [reply src/core/src/package_managers/ZypperPackageManager.py]: Done

File: src/core/src/bootstrap/EnvLayer.py
Comment (benank) [reply src/core/src/bootstrap/EnvLayer.py]: Got it, thanks. Fixed and defaulted to not raising.

File: src/core/src/bootstrap/EnvLayer.py
Comment (benank) [reply src/core/src/bootstrap/EnvLayer.py:60]: From what I have seen online, every Linux distro has its global environment variables stored here. But I understand your point and will keep it in mind in the future.

File: src/core/src/bootstrap/EnvLayer.py
Comment (rane-rajasi) [inline src/core/src/bootstrap/EnvLayer.py:105]: Do we need these 2 ifs?
First if finds any part in the file with var_name, could be a comment too while the second one looks for the exact pattern match, which is what we are looking for.

If this is right, then I think just the second if should be okay

Comment (benank) [reply src/core/src/bootstrap/EnvLayer.py:102]: Yep you're right, the first if would just save the execution of the regex logic. I'll remove the first if and leave it to the pattern matching.

File: src/core/src/bootstrap/EnvLayer.py
Comment (rane-rajasi) [inline src/core/src/bootstrap/EnvLayer.py:121]: Maybe in the next PR, if not enough time here, but the parts for writing when the env_var pre-exists in the the file and when it doesn't could be combined into one.

Refer https://github.com/Azure/LinuxPatchExtension/blob/aaa60bcb59f7a282ac8337a1cf31e9e120bb1e08/src/core/src/package_managers/AptitudePackageManager.py#L483 for a similar example where we write to a config file

File: src/core/src/bootstrap/EnvLayer.py
Comment (rane-rajasi) [inline src/core/src/bootstrap/EnvLayer.py:147]: Better to verify if env_var exists with a pattern match, as these files could contain comments with the variable name or the variable name in some other text format. Pattern match guarantees the part we fetch is exactly what we are looking for

File: src/core/tests/Test_CoreMain.py
Comment (rane-rajasi) [inline src/core/tests/Test_CoreMain.py:281]: Why did this value change?

File: src/core/src/package_managers/ZypperPackageManager.py
Comment (rane-rajasi) [inline src/core/src/package_managers/ZypperPackageManager.py:125]: Looks like we are depending on Constants.ERROR_ADDED_TO_STATUS to determine if retries are required, which does not look right.

 Constants.ERROR_ADDED_TO_STATUS is used to decide whether or not the error is to be added to error objects and should not be used for other cases. Better to have the usages independent, so that in future if the implementation for error objects change, it won't affect these retries.

File: src/core/src/package_managers/ZypperPackageManager.py
Comment (rane-rajasi) [inline src/core/src/package_managers/ZypperPackageManager.py:78]: Q: the error coming from command invocation, will be added to error objects from within package manager, either in line 110 or 132 in this file. Why do we need to add another error message?

Some context on error objects:
Error objects refers to the errors json within any summary substatus (i.e. PatchInstallationSummary or PatchAssesmentSummary, etc). Since it's added to a json file with possible size limitations, we only add contents actionable contents within it, which would help the user understand the cause of failure.

For eg: in this case, a message stating which command failed (from line 110 or 132). refresh_repo is a function defined in our code and adding a message "Unable to refresh repo (retries exhausted after reboot)" will not be of much help to the user. 

Add this error message to our logs but don't add it to the error objects, since we already have details on which command failed in there.

To do this, keep Constants.ERROR_ADDED_TO_STATUS in the error_message and simple use composite_logger (any log type you prefer), This will add the message to log files but not error objects in the status

Comment (benank) [reply src/core/src/package_managers/ZypperPackageManager.py]: I added this extra error message so that it the source of error could be extra visible - in this case, the error message could say that retries were exhausted after reboot. But I understand now what you mean with the extra context, and I will put the error message from this area into a warning log and keep the Constants.ERROR_ADDED_TO_STATUS in the message.

File: src/core/src/package_managers/ZypperPackageManager.py
Comment (benank) [reply src/core/src/package_managers/ZypperPackageManager.py:115]: Understood. I wanted some way to share the output code between the try/except blocks and opted for that. I have now just moved the code outside of those blocks so it's accessible to both.

File: src/core/tests/Test_CoreMain.py
Comment (benank) [reply src/core/tests/Test_CoreMain.py]: This value changed because of the retry logic. It was originally 2 (before any retries were added), but then set to 5 because of the 5 retries for refresh_repo and how it reported the errors. Now it is back to 2 because there are no more duplicate errors even with the new retries in the invocation.

File: src/core/src/bootstrap/EnvLayer.py
Comment (benank) [reply src/core/src/bootstrap/EnvLayer.py]: Thanks, done. Also removed the loop since it can just use the pattern patching.

File: src/core/src/bootstrap/EnvLayer.py
Comment (benank) [reply src/core/src/bootstrap/EnvLayer.py:121]: Got it, thanks.

Comment (rane-rajasi) [reply src/core/src/bootstrap/EnvLayer.py:121]: God improvement to the logic but consider having a backlog item to further improve this, for instance we could have a single write_with_retry for 

 if var_name not in environment_vars:
                self.file_system.write_with_retry(self.etc_environment_file_path, "\n" + formatted_env_var)
            else:
                # Update the value of the existing setting

File: src/core/src/package_managers/ZypperPackageManager.py
Comment (rane-rajasi) [inline src/core/src/package_managers/ZypperPackageManager.py:63]: minor: don't need return here, function should return if there is no exception and invoke_package_manager completes

**Review by rane-rajasi — APPROVED** [review]

File: src/core/src/package_managers/ZypperPackageManager.py
Comment (benank) [reply src/core/src/package_managers/ZypperPackageManager.py]: Right, fixed

**Pull request merged**
